### PR TITLE
[serve] Add exponential backoff when retrying replicas

### LINKED ISSUE #31121: [serve] Infinite restarting of replica when deployment constructor consistently fails
### What happened + What you expected to happen

Running a model that always fails upon initialization causes Serve to keep restarting the replica infinitely.
image

The deployment is marked unhealthy after 3 retries, which raises an exception in `serve.run()` (which by default waits for deployments to be healthy), which causes the script (or `serve run` CLI) to error out. The problem is, however, that this leaves Serve alive in the background, and Serve continues to try to reach its target state by restarting the replica infinitely.

### Versions / Dependencies

2.0

### Reproduction script

```
# test.py
from ray import serve

@serve.deployment
class A:
    def __init__(self):
        print("hi")
        1/0

app = A.bind()
```

```
# test.yaml
import_path: test.app
```
Run with `serve deploy test.yaml`

### Issue Severity

None

### PR DESCRIPTION
**Pull request by zcin**
Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

## Why are these changes needed?

If deployment is repeatedly failing, perform exponential backoff so as to not repeatedly try to restart the replica at a very fast rate.

## Related issue number

Closes #31121

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by sihanwang41 — APPROVED** [review]

File: python/ray/serve/_private/deployment_state.py [hunk 1]
```diff
@@ -978,6 +978,10 @@ def __init__(
         # DeploymentInfo and bring current deployment to meet new status.
         self._target_state: DeploymentTargetState = DeploymentTargetState.default()
         self._prev_startup_warning: float = time.time()
+        # Exponential backoff when retrying a consistently failing deployment
+        self._last_retry: float = 0.0
+        self._backoff_time: int = 1
+        self._max_backoff: int = 64
```
[the remark this answers is not in the document]
Comment (edoakes) [reply python/ray/serve/_private/deployment_state.py:984]: consider making these parametrizable via env var

**Review by edoakes — COMMENTED** [review]
Looks good.

@zcin what is the behavior under controller failure?

File: python/ray/serve/_private/deployment_state.py [hunk 2]
```diff
@@ -1308,28 +1313,43 @@ def _scale_deployment_replicas(self) -> bool:
             )
             to_add = max(delta_replicas - stopping_replicas, 0)
             if to_add > 0:
+                # Exponential backoff
+                failed_to_start_threshold = min(
+                    MAX_DEPLOYMENT_CONSTRUCTOR_RETRY_COUNT,
+                    self._target_state.num_replicas * 3,
```
Comment (architkulkarni) [inline python/ray/serve/_private/deployment_state.py:1319]: What's the reason for choosing `self._target_state.num_replicas * 3`?

Comment (architkulkarni) [reply python/ray/serve/_private/deployment_state.py:1319]: Ah is it to compare against the total number of replica restarts across the deployment? (So on average each replica has failed 3 times?)

Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1319]: I used the threshold used for setting the deployment unhealthy: [code pointer](https://github.com/ray-project/ray/blob/master/python/ray/serve/_private/deployment_state.py#L1388). Basically, perform exponential backoff after a replica fails 3 times and the deployment is determined unhealthy.

File: python/ray/serve/_private/deployment_state.py [hunk 3]
```diff
@@ -1308,28 +1313,43 @@ def _scale_deployment_replicas(self) -> bool:
             )
             to_add = max(delta_replicas - stopping_replicas, 0)
             if to_add > 0:
+                # Exponential backoff
+                failed_to_start_threshold = min(
```
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1319]: Synced offline.

File: python/ray/serve/_private/deployment_state.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py]: Thanks for the suggestions, applied!

File: python/ray/serve/_private/deployment_state.py (hunk 2 above)
Comment (zcin) [reply python/ray/serve/_private/deployment_state.py:1321]: > Ah is it to compare against the total number of replica restarts across the deployment? (So on average each replica has failed 3 times?)

Yup, I believe so.

**Comment by zcin** [comment]
> @zcin what is the behavior under controller failure?

@edoakes Not sure if there's a way to test, but I believe all related state is reset. From what I read in the code, `_replica_constructor_retry_counter` itself is reset, and `backoff_time_s` is as well; the only state recovered for each deployment is the target state.

**Review by architkulkarni — APPROVED** [review]
Looks good!

**Pull request merged**